Core/src/IfcGeometryConverter/CurveConverter.h

```diff
@@ -815,6 +769,143 @@ namespace OpenInfraPlatform {
+					const typename IfcEntityTypesT::IfcSegmentIndexSelect & segmentIndexSelect,
+					const std::vector<carve::geom::vector<3>>& points)const throw(...)
+				{
+					switch (segmentIndexSelect.which())
+					{
+					case 0:
+					{
+						// TYPE IfcArcIndex = LIST [3:3] OF IfcPositiveInteger;
+						return convertIfcArcIndex(segmentIndexSelect.get<0>(), points);
+					}
+					case 1:
+					{
+						// TYPE IfcLineIndex = LIST [2:?] OF IfcPositiveInteger;
+						return convertIfcLineIndex(segmentIndexSelect.get<1>(), points);
+					}
+					default:
+						throw oip::UnhandledException();
+					}
+				}
+				
+				std::vector<carve::geom::vector<3>> convertIfcLineIndex(
+					const typename IfcEntityTypesT::IfcLineIndex &arcSegment,
+					const std::vector<carve::geom::vector<3>> points) const throw(...)
+				{
+					if (arcSegment.size() < 2)
+						throw oip::InconsistentModellingException("The number of indices for one of IfcLineIndex is less than 2!");
+
+					std::vector<carve::geom::vector<3>> loop;
+					for (const auto& i : arcSegment)
+					{
+						loop.push_back(points[i - 1]); //EXPRESS count from 1, C++ from 0
+					}
+					return loop;
+				}
+
+
+				
+				std::vector<carve::geom::vector<3>> convertIfcArcIndex(
+					const typename IfcEntityTypesT::IfcArcIndex &arcSegment, 
+					const std::vector<carve::geom::vector<3>> points) const throw(...)
+				{
+					if (arcSegment.size() != 3)
+						throw oip::InconsistentModellingException("The number of indices for one of IfcArcIndex is not 3!");
+
+					carve::geom::vector<3> arcStart = points[arcSegment[0] - 1];
+					carve::geom::vector<3> arcMid = points[arcSegment[1] - 1];
+					carve::geom::vector<3> arcEnd = points[arcSegment[2] - 1];
+					
+					//Converting a 3D to a 2D problem to then find the circle from three points on the plane.
+					// see https://math.stackexchange.com/questions/1379217/3-points-in-3d-space-to-find-the-center-of-an-arc-or-circle
+
+					//Start by finding the normal vector to the plane defined by the three points
+					//n=unitvector(A×B+B×C+C×A)
+					carve::geom::vector<3> normalVector = carve::geom::cross(arcStart, arcMid) + carve::geom::cross(arcMid, arcEnd) + carve::geom::cross(arcEnd, arcStart);
+					normalVector.normalize();
+					//u=(C−A)×n
+					carve::geom::vector<3> firstOrthogonalDirection = carve::geom::cross(arcEnd - arcStart, normalVector);
+					firstOrthogonalDirection.normalize();
+					//v=n×u
+					carve::geom::vector<3> secondOrthogonalDirection = carve::geom::cross(normalVector, firstOrthogonalDirection);
+					secondOrthogonalDirection.normalize();
+		
+					//Calculate distance of the plane to the origin
+					double distance = (normalVector.x*arcStart.x) + (normalVector.y*arcStart.y) + (normalVector.y*arcStart.y);
+
+					//Convert the problem into a 2D problem
+					carve::math::Matrix rotationMatrix = carve::math::Matrix(
+						firstOrthogonalDirection.x, secondOrthogonalDirection.x, 0, 0,

```

Comment:
```suggestion
						firstOrthogonalDirection.x, secondOrthogonalDirection.x, 0., 0.,
```
And similar to all entries to the matrix.

Reply:
Resolved: f9547d9